packages/mobile/src/app/saga.ts

```diff
@@ -144,39 +156,75 @@ export function* handleDeepLink(action: OpenDeepLink) {
   }
 }
 
-export function* navigatePinProtected(action: NavigatePinProtected) {
+export function* handleNavigatePinProtected(action: NavigatePinProtected) {
   const fornoMode = yield select(fornoSelector)
   try {
+    // TODO: Implement PIN protection for forno mode
     if (!fornoMode) {
-      const pincode = yield call(getPincode, false)
-      yield put(startPinVerification())
-      const account = yield call(getAccount)
-      yield call(web3.eth.personal.unlockAccount, account, pincode, UNLOCK_DURATION)
-      navigate(action.routeName, action.params)
-      yield put(finishPinVerification())
-    } else {
-      // TODO: Implement PIN protection for forno mode
-      navigate(action.routeName, action.params)
+      yield call(getPincode, false)
     }
+    navigate(action.routeName, action.params)
   } catch (error) {
     Logger.error(TAG + '@showBackupAndRecovery', 'Incorrect pincode', error)
     yield put(showError(ErrorMessages.INCORRECT_PIN))
-    yield put(finishPinVerification())
   }
 }
 
 export function* watchNavigatePinProtected() {
-  yield takeLatest(Actions.NAVIGATE_PIN_PROTECTED, navigatePinProtected)
+  yield takeLatest(Actions.NAVIGATE_PIN_PROTECTED, handleNavigatePinProtected)
 }
 
 export function* watchDeepLinks() {
   yield takeLatest(Actions.OPEN_DEEP_LINK, handleDeepLink)
 }
 
+function createAppStateChannel() {
+  return eventChannel((emit: any) => {
+    AppState.addEventListener('change', emit)
+
+    const removeEventListener = () => {
+      AppState.removeEventListener('change', emit)
+    }
+    return removeEventListener
+  })
+}
+
+function* watchAppState() {
+  Logger.debug(`${TAG}@monitorAppState`, 'Starting monitor app state saga')
+  const appStateChannel = yield createAppStateChannel()
+  while (true) {
+    try {
+      const newState = yield take(appStateChannel)
+      Logger.debug(`${TAG}@monitorAppState`, `App changed state: ${newState}`)
+      yield put(setAppState(newState))
+    } catch (error) {
+      Logger.error(`${TAG}@monitorAppState`, error)
+    } finally {
+      if (yield cancelled()) {
+        appStateChannel.close()
+      }
+    }
+  }
+}
+
+function* handleSetAppState(action: SetAppState) {
+  const appLocked = yield select(getAppLocked)
+  const cachedPin = getCachedPincode()
+  const lockWithPinEnabled = yield select(getLockWithPinEnabled)
+  if (lockWithPinEnabled && action.state === 'background' && !appLocked && !cachedPin) {
+    navigate(Screens.Background, {
+      onUnlock() {
+        navigateBack({ immediate: true })
+      },
+    })
+  }
+}
+
 export function* appSaga() {
   yield spawn(navigateToProperScreen)
   yield spawn(toggleToProperSyncMode)
-  yield spawn(checkAppDeprecation)
   yield spawn(watchNavigatePinProtected)
   yield spawn(watchDeepLinks)
+  yield spawn(watchAppState)
+  yield takeLatest(Actions.SET_APP_STATE, handleSetAppState)
```

Comment:
instead of watching for SET_APP_STATE couldn't we just do the handleSetAppState in watchAppState where we're already taking from the event channel?

Reply:
I think of it like this: one place to emit `Actions.SET_APP_STATE` and other places are subscribed to this event (reducer, saga(s)). This way other modules, like `geth`, could just subscribe to `Actions.SET_APP_STATE` too if they need it.  In this implementation `watchAppState` is rather a propagator than a handler, if that makes sense.